src/Memory/InputQueue.ts

```diff
@@ -6,117 +6,149 @@ import * as BB from 'botbuilder'
+        if (!this.inputQueues[conversationId]) {
+            this.inputQueues[conversationId] = []
+        }
+        this.inputQueues[conversationId].push(queuedInput)
+        this.log(`ADD QUEUE`, conversationId, queuedInput.activityId)
     }
 
-    // Attempt to process next message in the queue
-    private static async InputQueueProcess(inProcessMessageState: InProcessMessageState): Promise<void> {
+    private static hasExpired(queuedInput: QueuedInput): boolean {
         const now = new Date().getTime()
-        const messageProcessing = await inProcessMessageState.get<QueuedInput>()
+        const age = now - queuedInput.timestamp
+        return (age > MESSAGE_TIMEOUT)
+    }
+
+    // Process next input
+    private static async InputQueueProcessNext(inProcessMessageState: InProcessMessageState, conversationId: string): Promise<void> {
 
-        // Is a message being processed
-        if (messageProcessing) {
+        // Get current input being processed (mutex)
+        let inputInProcess = await inProcessMessageState.get<QueuedInput>()
 
-            // Remove if it's been expired
-            const age = now - messageProcessing.timestamp
+        // If input is being processed (mutex is set), check if it has expired
+        if (inputInProcess) {
+            if (this.hasExpired(inputInProcess)) {
+                // Item in mutex has expired
+                this.log(`EXPIRED U`, conversationId, inputInProcess.activityId)
 
-            if (age > MESSAGE_TIMEOUT) {
-                CLDebug.Log(`EXPIRED: ${messageProcessing.activityId} ${this.messageQueue.length}`, DebugType.MessageQueue)
+                // Clear the input mutex
                 await inProcessMessageState.remove()
-                let queuedInput = this.messageQueue.find(mq => mq.activityId == messageProcessing.activityId)
-                if (queuedInput) {
-                    // Fire the callback with failure
-                    queuedInput.callback(true, queuedInput.activityId)
-                }
-                else {
-                    CLDebug.Log(`EXPIRE-WARNING: Couldn't find queued message`, DebugType.MessageQueue)
-                }
-                CLDebug.Log(`EXPIRE-POP: ${messageProcessing.activityId} ${this.messageQueue.length}`, DebugType.MessageQueue)
-            }
-        }
 
-        // If no message being processed, try next message
-        if (!messageProcessing) {
-            await InputQueue.InputQueueProcessNext(inProcessMessageState)
+                // Process the next input
+                await InputQueue.InputQueueProcessNext(inProcessMessageState, conversationId)
+            }
         }
-    }
-
-    // Process next message
-    private static async InputQueueProcessNext(inProcessMessageState: InProcessMessageState): Promise<void> {
+        // Otherwise process the next one
+        else if (this.inputQueues[conversationId]) {
+            let inputToProcess = this.inputQueues[conversationId].shift()
```

Comment:
Hard to tell from the diff view, did this have to be `let`?  I expected this to not be reassigned, it's only dequeuing